File: snake.c
#include <ncurses.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <stdbool.h>

// 游戏区域大小
#define HEIGHT 20
#define WIDTH 40

// 方向键
#define UP_KEY     KEY_UP
#define DOWN_KEY   KEY_DOWN
#define LEFT_KEY   KEY_LEFT
#define RIGHT_KEY  KEY_RIGHT

// 游戏状态
typedef struct {
    int x, y;    // 坐标
} Position;

// 蛇的结构
typedef struct {
    Position pos[100];  // 蛇身坐标数组
    int length;         // 当前长度
    int direction;      // 移动方向
} Snake;

// 食物结构
typedef struct {
    Position pos;       // 食物位置
} Food;

// 游戏状态
typedef struct {
    Snake snake;
    Food food;
    int score;
    bool running;
} Game;

// 初始化游戏
void init_game(Game* game) {
    // 初始化随机数
    srand(time(NULL));
    
    // 初始化蛇
    game->snake.length = 3;
    game->snake.direction = RIGHT_KEY;
    game->snake.pos[0].x = WIDTH/2;
    game->snake.pos[0].y = HEIGHT/2;
    game->snake.pos[1].x = WIDTH/2-1;
    game->snake.pos[1].y = HEIGHT/2;
    game->snake.pos[2].x = WIDTH/2-2;
    game->snake.pos[2].y = HEIGHT/2;
    
    // 初始化食物
    game->food.pos.x = rand() % (WIDTH-2) + 1;
    game->food.pos.y = rand() % (HEIGHT-2) + 1;
    
    // 初始化分数
    game->score = 0;
    game->running = true;
}

// 生成新的食物
void new_food(Game* game) {
    bool valid;
    do {
        valid = true;
        game->food.pos.x = rand() % (WIDTH-2) + 1;
        game->food.pos.y = rand() % (HEIGHT-2) + 1;
        
        // 确保食物不会出现在蛇身上
        for(int i = 0; i < game->snake.length; i++) {
            if(game->food.pos.x == game->snake.pos[i].x && 
               game->food.pos.y == game->snake.pos[i].y) {
                valid = false;
                break;
            }
        }
    } while(!valid);
}

// 绘制游戏界面
void draw(Game* game) {
    // 清屏
    clear();
    
    // 绘制边框
    for(int i = 0; i < HEIGHT; i++) {
        mvaddch(i, 0, '|');
        mvaddch(i, WIDTH-1, '|');
    }
    for(int i = 0; i < WIDTH; i++) {
        mvaddch(0, i, '-');
        mvaddch(HEIGHT-1, i, '-');
    }
    
    // 绘制蛇
    for(int i = 0; i < game->snake.length; i++) {
        if(i == 0) {
            // 蛇头
            attron(COLOR_PAIR(1));
            mvaddch(game->snake.pos[i].y, game->snake.pos[i].x, '@');
            attroff(COLOR_PAIR(1));
        } else {
            // 蛇身
            attron(COLOR_PAIR(2));
            mvaddch(game->snake.pos[i].y, game->snake.pos[i].x, 'o');
            attroff(COLOR_PAIR(2));
        }
    }
    
    // 绘制食物
    attron(COLOR_PAIR(3));
    mvaddch(game->food.pos.y, game->food.pos.x, '*');
    attroff(COLOR_PAIR(3));
    
    // 显示分数
    mvprintw(0, WIDTH/2-5, "Score: %d", game->score);
    
    // 刷新屏幕
    refresh();
}

// 移动蛇
void move_snake(Game* game) {
    // 保存蛇头位置
    Position new_head = game->snake.pos[0];
    
    // 根据方向移动蛇头
    switch(game->snake.direction) {
        case UP_KEY:
            new_head.y--;
            break;
        case DOWN_KEY:
            new_head.y++;
            break;
        case LEFT_KEY:
            new_head.x--;
            break;
        case RIGHT_KEY:
            new_head.x++;
            break;
    }
    
    // 检查是否撞墙
    if(new_head.x <= 0 || new_head.x >= WIDTH-1 ||
       new_head.y <= 0 || new_head.y >= HEIGHT-1) {
        game->running = false;
        return;
    }
    
    // 检查是否撞到自己
    for(int i = 0; i < game->snake.length; i++) {
        if(new_head.x == game->snake.pos[i].x && 
           new_head.y == game->snake.pos[i].y) {
            game->running = false;
            return;
        }
    }
    
    // 移动蛇身
    for(int i = game->snake.length-1; i > 0; i--) {
        game->snake.pos[i] = game->snake.pos[i-1];
    }
    game->snake.pos[0] = new_head;
    
    // 检查是否吃到食物
    if(new_head.x == game->food.pos.x && new_head.y == game->food.pos.y) {
        game->score++;
        game->snake.length++;
        new_food(game);
    }
}

int main() {
    // 初始化ncurses
    initscr();
    noecho();
    curs_set(0);
    keypad(stdscr, TRUE);
    timeout(100);
    
    // 初始化颜色
    start_color();
    init_pair(1, COLOR_RED, COLOR_BLACK);    // 蛇头
    init_pair(2, COLOR_GREEN, COLOR_BLACK);  // 蛇身
    init_pair(3, COLOR_YELLOW, COLOR_BLACK); // 食物
    
    // 初始化游戏
    Game game;
    init_game(&game);
    
    // 主游戏循环
    while(game.running) {
        // 处理输入
        int key = getch();
        switch(key) {
            case UP_KEY:
                if(game.snake.direction != DOWN_KEY)
                    game.snake.direction = UP_KEY;
                break;
            case DOWN_KEY:
                if(game.snake.direction != UP_KEY)
                    game.snake.direction = DOWN_KEY;
                break;
            case LEFT_KEY:
                if(game.snake.direction != RIGHT_KEY)
                    game.snake.direction = LEFT_KEY;
                break;
            case RIGHT_KEY:
                if(game.snake.direction != LEFT_KEY)
                    game.snake.direction = RIGHT_KEY;
                break;
            case 'q':
                game.running = false;
                break;
        }
        
        // 移动蛇
        move_snake(&game);
        
        // 绘制游戏
        draw(&game);
        
        // 控制游戏速度
        usleep(100000);  // 100ms
    }
    
    // 游戏结束
    mvprintw(HEIGHT/2, WIDTH/2-10, "Game Over! Score: %d", game.score);
    mvprintw(HEIGHT/2+1, WIDTH/2-15, "Press any key to exit...");
    refresh();
    getch();
    
    // 清理
    endwin();
    return 0;
} 

File: snake.py
#!/usr/bin/env python3
import curses
import random
import time

# 游戏设置
HEIGHT = 20
WIDTH = 40
SPEED = 0.1  # 游戏速度（秒）

class SnakeGame:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        self.height = HEIGHT
        self.width = WIDTH
        
        # 初始化颜色
        curses.start_color()
        curses.init_pair(1, curses.COLOR_RED, curses.COLOR_BLACK)     # 蛇头
        curses.init_pair(2, curses.COLOR_GREEN, curses.COLOR_BLACK)   # 蛇身
        curses.init_pair(3, curses.COLOR_YELLOW, curses.COLOR_BLACK)  # 食物
        
        # 初始化游戏窗口
        self.win = curses.newwin(self.height + 2, self.width + 2, 0, 0)
        self.win.keypad(1)
        self.win.timeout(100)
        
        # 初始化游戏状态
        self.reset_game()
    
    def reset_game(self):
        """重置游戏状态"""
        # 初始化蛇的位置（从中间开始）
        self.snake = [
            [self.height//2, self.width//2],      # 蛇头
            [self.height//2, self.width//2 - 1],  # 蛇身
            [self.height//2, self.width//2 - 2]   # 蛇尾
        ]
        self.direction = curses.KEY_RIGHT
        self.score = 0
        self.food = self.generate_food()
        self.game_over = False
    
    def generate_food(self):
        """生成新的食物"""
        while True:
            food = [
                random.randint(1, self.height-2),
                random.randint(1, self.width-2)
            ]
            if food not in self.snake:
                return food
    
    def draw(self):
        """绘制游戏界面"""
        # 清空窗口
        self.win.clear()
        
        # 绘制边框
        self.win.border(0)
        
        # 绘制蛇
        for i, (y, x) in enumerate(self.snake):
            if i == 0:  # 蛇头
                self.win.addstr(y, x, '@', curses.color_pair(1))
            else:       # 蛇身
                self.win.addstr(y, x, 'o', curses.color_pair(2))
        
        # 绘制食物
        self.win.addstr(self.food[0], self.food[1], '*', curses.color_pair(3))
        
        # 显示分数
        score_text = f"Score: {self.score}"
        self.win.addstr(0, (self.width - len(score_text))//2, 
                       score_text, curses.color_pair(3))
        
        # 刷新窗口
        self.win.refresh()
    
    def move(self):
        """移动蛇"""
        # 获取蛇头位置
        head = self.snake[0].copy()
        
        # 根据方向移动蛇头
        if self.direction == curses.KEY_UP:
            head[0] -= 1
        elif self.direction == curses.KEY_DOWN:
            head[0] += 1
        elif self.direction == curses.KEY_LEFT:
            head[1] -= 1
        elif self.direction == curses.KEY_RIGHT:
            head[1] += 1
        
        # 检查是否撞墙
        if (head[0] <= 0 or head[0] >= self.height-1 or
            head[1] <= 0 or head[1] >= self.width-1):
            self.game_over = True
            return
        
        # 检查是否撞到自己
        if head in self.snake:
            self.game_over = True
            return
        
        # 移动蛇
        self.snake.insert(0, head)
        
        # 检查是否吃到食物
        if head == self.food:
            self.score += 1
            self.food = self.generate_food()
        else:
            self.snake.pop()
    
    def run(self):
        """运行游戏主循环"""
        while not self.game_over:
            # 处理输入
            key = self.win.getch()
            if key == -1:
                pass
            elif key in [curses.KEY_UP, curses.KEY_DOWN, 
                        curses.KEY_LEFT, curses.KEY_RIGHT]:
                # 防止反向移动
                if (key == curses.KEY_UP and self.direction != curses.KEY_DOWN) or \
                   (key == curses.KEY_DOWN and self.direction != curses.KEY_UP) or \
                   (key == curses.KEY_LEFT and self.direction != curses.KEY_RIGHT) or \
                   (key == curses.KEY_RIGHT and self.direction != curses.KEY_LEFT):
                    self.direction = key
            elif key == ord('q'):  # 按q退出
                break
            
            # 移动蛇
            self.move()
            
            # 绘制游戏
            self.draw()
            
            # 控制游戏速度
            time.sleep(SPEED)
        
        # 游戏结束显示
        if self.game_over:
            game_over_text = f"Game Over! Score: {self.score}"
            self.win.addstr(self.height//2, 
                           (self.width - len(game_over_text))//2,
                           game_over_text)
            self.win.addstr(self.height//2 + 1, 
                           (self.width - 20)//2,
                           "Press any key to exit...")
            self.win.refresh()
            self.win.getch()

def main(stdscr):
    """主函数"""
    # 设置
    curses.curs_set(0)  # 隐藏光标
    stdscr.clear()
    
    # 创建并运行游戏
    game = SnakeGame(stdscr)
    game.run()

if __name__ == "__main__":
    # 使用 curses.wrapper 来正确处理终端设置
    curses.wrapper(main) 